### PR TITLE
Completed Screwtape project

### PR DESCRIPTION
**Pull request by HumaGitGud**

### CODE REVIEW

File: src/Node.java [hunk 1]
```diff
@@ -38,7 +39,23 @@ public Node(int value) {
...
+    // set current node
+    Node curr = this;
+
+    // fill elements of list in new nodes with references
+    for (int i = 1; i < list.size(); i++) {
+      curr.next = new Node(list.get(i));
+      curr.next.prev = curr;
+
+      curr = curr.next;
+    }
```
Comment (auberonedu) [inline src/Node.java:58]: Nice logic!

File: src/NodeTest.java [hunk 2]
```diff
@@ -42,9 +42,22 @@ void testListConstructorWithEmptyList() {
   }
 
   
-  // TODO: Add test for list constructor when passed null list
-  // TODO: Add at least one more test for list constructor that would be useful and cover new ground.
+  // Test for list constructor when passed null list
+  @Test
+  void testListConstructorWithNullList() {
+    assertThrows(IllegalArgumentException.class, () -> new Node(null), "Cannot pass null as a list");
+  }
```
Comment (auberonedu) [inline src/NodeTest.java:49]: Nice concise test!

File: src/ScrewtapeInterpreter.java [hunk 3]
```diff
@@ -105,9 +107,34 @@ public void moveTapePointerToTail() {
...
+        stack.push(i);
+      } else if (c == ']') { // if its ] it means a match is found for [
+        if (stack.isEmpty()) {
+          throw new IllegalArgumentException("Program contains unmatched closing bracket at index " + i);
+        }
+        int open = stack.pop(); // save index coordinate before popping it
+        map.put(i, open); // record index of ] with the popped coordinate of [ open
+      } 
+    }
+
```
Comment (auberonedu) [inline src/ScrewtapeInterpreter.java:131]: Nice logic!

File: src/ScrewtapeInterpreter.java [hunk 4]
```diff
@@ -129,8 +156,43 @@ public Map<Integer, Integer> bracketMap(String program) {
...
+          tapePointer.next.prev = tapePointer; // link prev
+        }
+        tapePointer = tapePointer.next; // advance node forward
+      } else if (c == '<') {
+        if (tapePointer.prev == null) {
+          tapePointer.prev = new Node(0);
+          tapeHead = tapePointer.prev; // set new head
+          tapeHead.next = tapePointer;
+        }
+        tapePointer = tapePointer.prev;
```
Comment (auberonedu) [inline src/ScrewtapeInterpreter.java:182]: Nice!

File: src/ScrewtapeInterpreter.java [hunk 5]
```diff
@@ -129,8 +156,43 @@ public Map<Integer, Integer> bracketMap(String program) {
    * @throws IllegalArgumentException If the program contains unmatched brackets.
    */
   public String execute(String program) {
-    // TODO: Implement this
-    // If you get stuck, you can look at hint.md for a hint
-    return null;
+    Map<Integer, Integer> map = bracketMap(program); // use the map of brackets to jump back/ loop
+    String output = "";
```
Comment (auberonedu) [inline src/ScrewtapeInterpreter.java:160]: Consider using StringBuilder when repeatedly adding to a String.

File: src/ScrewtapeInterpreterTest.java [hunk 6]
```diff
@@ -23,8 +23,71 @@ void testNestedBracketMap() {
...
+  @Test
+  void testBracketMapInvalidProgram() {
+    ScrewtapeInterpreter interpreter = new ScrewtapeInterpreter();
+    String extraOpen = "[[[]";
+    String extraClose = "]][]";
+    String extraMix = "][[]";
+    assertThrows(IllegalArgumentException.class, () -> interpreter.bracketMap(extraOpen));
+    assertThrows(IllegalArgumentException.class, () -> interpreter.bracketMap(extraClose));
+    assertThrows(IllegalArgumentException.class, () -> interpreter.bracketMap(extraMix));
+  }
```
Comment (auberonedu) [inline src/ScrewtapeInterpreterTest.java:90]: Nice tests!

**Review by auberonedu — COMMENTED** [review]
Great job!
